### DISCUSSION TITLE
Custom frame preprocessing

### DISCUSSION THREAD

**Discussion by wasilukm** [pymodbus-dev/pymodbus#2476] [Help] [answered]
Hello,

I have a device that incorrectly creates Modbus frames - the data length is incorrectly calculated. Therefore before any further frame processing by pymodbus, the frame has to be manipulated to correct the data length, otherwise, there is a justified exception "RuntimeError: Received error response Modbus Error: [Input/Output] Unable to decode request"

I was using a custom framer below PyModbus 3.7. In PyModbus 3.7.2, I could still use the custom framer approach but with the "old framer" API. In 3.7.3, the old framer was removed, so I switched to a custom Decoder. However, in 3.7.4, Decoders were reworked, so again I had to update my code. As I understand Decoder is treated as an internal interface so it can change in any release. Therefore, I'm wondering, if it is possible to accomplish the same with only the official API.

Currently, my code is like this:

```
from pymodbus.client import ModbusTcpClient
from pymodbus.pdu import DecodePDU


class CustomDecoder(DecodePDU):
    
    def decode(self, frame):
        # fix frame here
        return super().decode(frame)

client = ModbusTcpClient('1.2.3.4')

client.framer.decoder = CustomDecoder(is_server=False)
```

**Comment by janiversen** [comment, marked as answer]
With the upcoming v3.8.0 (on dev) there are trace points where you can add your own methods and manipulate the data

There are 4 trace points:
- receive/send byte stream
- receive/send pdu

Currently we have not published an API to attach them, that is due later this week.

**Reply by wasilukm** [reply]
Great , so I'm looking forward for 3.8.0 features!
Thank you for prompt response.

**Comment by janiversen** [comment]
Actually the most proper solution would be to register a custom pdu, which overwrites the standard pdu, that way you can "persuade" pymodbus to accept a non-standard packet. The custom pdu can simply inherit from the standard pdu, and only overwrite (probably decode()) what is needed.

**Reply by wasilukm** [reply]
Nice, indeed it is working.

I registered a custom ReadHoldingRegistersResponse class and got it working in all 3.7 versions despite the interface being considered internal, so I don't have to require the exact pymodbus version.

Thank you.
